Client/Frontend/Browser/TabTrayController.swift

```diff
@@ -897,3 +909,54 @@ private class EmptyPrivateTabsView: UIView {
         fatalError("init(coder:) has not been implemented")
     }
 }
+
+@available(iOS 9.0, *)
+extension TabTrayController: UIViewControllerPreviewingDelegate {
+
+    func previewingContext(previewingContext: UIViewControllerPreviewing, viewControllerForLocation location: CGPoint) -> UIViewController? {
+
+        guard let collectionView = collectionView else { return nil }
+        let convertedLocation = self.view.convertPoint(location, toView: collectionView)
+
+        guard let indexPath = collectionView.indexPathForItemAtPoint(convertedLocation),
+            let cell = collectionView.cellForItemAtIndexPath(indexPath) else { return nil }
+
+        let tab = tabDataSource.tabs[indexPath.row]
+        let tabVC = TabPeekViewController(tab: tab, controller: browserViewController, tabManager: tabManager)
+        if let browserProfile = profile as? BrowserProfile {
+            tabVC.setState(withProfile: browserProfile, clientPickerDelegate: self)
+        }
+        previewingContext.sourceRect = cell.frame
+
+        return tabVC
+    }
+
+    func previewingContext(previewingContext: UIViewControllerPreviewing, commitViewController viewControllerToCommit: UIViewController) {
+        guard let tpvc = viewControllerToCommit as? TabPeekViewController else { return }
+        tabManager.selectTab(tpvc.tab)
+        self.navigationController?.popViewControllerAnimated(true)
+        browserViewController?.urlBar.updateAlphaForSubviews(1)
+
+        [browserViewController?.header,
+            browserViewController?.footer,
+            browserViewController?.readerModeBar,
+            browserViewController?.footerBackdrop,
+            browserViewController?.headerBackdrop].forEach { view in
+            view?.transform = CGAffineTransformIdentity
```

Comment:
Can you include a comment here? Not really obvious what this code is doing, or why it's needed.